controller/src/main/java/io/pravega/controller/store/stream/InMemoryStream.java

```diff
@@ -9,515 +9,405 @@
+        configuration = create.getConfiguration();
         return CompletableFuture.completedFuture(null);
     }
 
     @Override
-    public String getScope() {
-        return this.scopeName;
+    CompletableFuture<Void> setConfigurationData(StreamConfiguration configuration) {
+        this.configuration = configuration;
+        return CompletableFuture.completedFuture(null);
     }
 
     @Override
-    public String getName() {
-        return this.streamName;
+    CompletableFuture<StreamConfiguration> getConfigurationData() {
+        if (this.configuration == null) {
+            return FutureHelpers.failedFuture(new DataNotFoundException(getName()));
+        }
+        return CompletableFuture.completedFuture(this.configuration);
     }
 
     @Override
-    public String getScopeName() {
-        return this.scopeName;
+    CompletableFuture<Void> createState(State state) {
+        this.state.set(new Data<>(SerializationUtils.serialize(state), 0));
+        return CompletableFuture.completedFuture(null);
     }
 
     @Override
-    public synchronized CompletableFuture<Boolean> updateConfiguration(StreamConfiguration configuration) {
-        this.configuration = configuration;
-        return CompletableFuture.completedFuture(true);
+    CompletableFuture<Void> setStateData(Data<Integer> newState) {
+        CompletableFuture<Void> result = new CompletableFuture<>();
+        this.state.updateAndGet(x -> {
+            if (Objects.equals(x.getVersion(), newState.getVersion())) {
+                result.complete(null);
+                return new Data<>(newState.getData(), newState.getVersion() + 1);
+            } else {
+                result.completeExceptionally(new WriteConflictException("state"));
+                return x;
+            }
+        });
+
+        return result;
     }
 
     @Override
-    public synchronized CompletableFuture<StreamConfiguration> getConfiguration() {
-        return CompletableFuture.completedFuture(this.configuration);
+    CompletableFuture<Data<Integer>> getStateData() {
+        if (this.state.get() == null) {
+            return FutureHelpers.failedFuture(new DataNotFoundException(getName()));
+        }
+
+        return CompletableFuture.completedFuture(state.get());
     }
 
     @Override
-    public CompletableFuture<Boolean> updateState(State state) {
-        this.state = state;
-        return CompletableFuture.completedFuture(true);
+    CompletableFuture<Void> createSegmentTable(Create create) {
+        final int numSegments = create.getConfiguration().getScalingPolicy().getMinNumSegments();
+        final double keyRangeChunk = 1.0 / numSegments;
+
+        final int startingSegmentNumber = 0;
+        final List<AbstractMap.SimpleEntry<Double, Double>> newRanges = IntStream.range(0, numSegments)
+                .boxed()
+                .map(x -> new AbstractMap.SimpleEntry<>(x * keyRangeChunk, (x + 1) * keyRangeChunk))
```

Comment:
Yes, and it would be a lot cleaner if those methods took a Map<Double, Double> or a SortedMap<Double,Double>.